cpp/src/arrow/compute/exec/options.h

```diff
@@ -430,6 +430,23 @@ class ARROW_EXPORT SelectKSinkNodeOptions : public SinkNodeOptions {
   SelectKOptions select_k_options;
 };
 
+/// \brief Make a node which selects a range of rows passed through it
+///
+/// All batches pushed to this node will be accumulated, then selected, by the given
```

Comment:
If I just want to select a range of rows, why accumulate all batches? In R how we've implemented `head()` now is to kill the plan after we've collected enough rows: https://github.com/apache/arrow/blob/master/r/R/query-engine.R#L241

Reply:
We need to optimize this further. I created separate [ticket](https://issues.apache.org/jira/browse/ARROW-17196) for that. This effort is mainly to support the Substrait integration.

Reply:
We already have [ARROW-14202](https://issues.apache.org/jira/browse/ARROW-14202) which is pretty similar to ARROW-17196.  I'm not sure if a new JIRA is needed. CC @ArianaVillegas who has done some initial thinking about ARROW-14202 and was thinking about implementing it at  some point.